Feature: select all fields from a table

hotaru355 opened the pull request:
## Problem

We would like to be able to select all fields (excluding any relations!) from a table. This PR also adds a refactor to the `TableModel` which prepares us for allowing any other model libraries to be used as the table models.

## Changes

- added `invert_selection` argument to all `returning`/`yielding`/... methods to list all fields that should **not** be included. Using `invert_selection=True` (default=`False`) and `columns=[]` (default=`[]`), you will get all columns.
- refactored `TableModel` so all pydantic-specific code is encapsulated in the `PydanticV1TableModel` class and implements the abstract interface in the `TableModel` class. any other library such as pydantic V2 and msgspec that implements this interface will be usable as a table model.
- replace `black` with `ruff` (responsible for most line changes, SORRY! non-formatting changes are marked with comments)
- Tests now run locally, not in a container (!). This is handy for debugging test cases. The database and Hasura Server are still containerized.
- `pytest-asyncio` bug that prevented session-scoped event loops is now fixed and so workaround is removed and annotation added to all async test classes
- small refactorings

## Steps to Test or Reproduce

- tests ensure there are no breaking changes
- added new test case for this feature
- you can test this locally by spinning up the containers

Commit:
wip

hotaru355 committed:
change to run tests not in docker, ruff instead of black

hotaru355 committed:
wip

hotaru355 committed:
wip

hotaru355 committed:
Merge branch 'feature/invert-column-selection' of github.com:hotaru355/cuckoo-hasura into feature/invert-column-selection

hotaru355 committed:
tests fixed

hotaru355 commented on .env.default:
```diff
@@ -1,4 +1,5 @@
-HASURA_URL=http://hasura:8080/v1/graphql
+HASURA_PORT=8080
+HASURA_HEALTH_URL="http://localhost:${HASURA_PORT}/healthz"
```
pulled out some const values, but these are not required in any `.env` file, as they have default values

hotaru355 commented on .github/workflows/ci_tests.yml:
```diff
@@ -15,25 +15,28 @@ jobs:
         uses: actions/checkout@v3
         with:
           path: ./
-      - name: Build test container
```
OLD:
- build not only the DB and Hasura container, but also a container to run the tests
- scp the coverage report out of the container to publish it on github

hotaru355 commented on .github/workflows/ci_tests.yml:
```diff
@@ -15,25 +15,28 @@ jobs:
         with:
           path: ./
-      - name: Build test container
+      - name: Start Database and Hasura Server
```
NEW:
- only start DB and Hasura container
- run tests locally

hotaru355 commented on cuckoo/finalizers.py:
```diff
@@ -61,6 +63,20 @@ def __init__(
         super().__init__()
         self._node: BinaryTreeNode = node
 
+    def _resolve_column_selection(
```
This private method is in essence the new feature. it is called from all methods that have a `columns` argument. Note that the default value for `columns` is now dynamic: depending on the `invert_selection` the defaults are either `["uuid"]` or `[]`. Let me know if you think that is confusing.

hotaru355 commented on cuckoo/finalizers.py:
```diff
@@ -61,6 +63,20 @@ def __init__(
+        if columns is None:
+            columns = DEFAULT_COLUMNS_INVERTED if invert_selection else DEFAULT_COLUMNS
+        if invert_selection:
+            columns_sanatized = {col for col in columns if isinstance(col, str)}
```
if you invert the selection, `Include`s are invalid items in the `columns` list and are silently ignored.

kaymanb replied:
I would prefer if an error was raised, instead of ignoring `Include` statements. 

Also, how does this work if I pass in `columns=["author { name, address }"]` instead of using an `Include`?

clvcooke replied:
Ya I use the columns as a string hack sometimes

hotaru355 replied:
good point! will raise an error instead

hotaru355 commented on cuckoo/finalizers.py:
```diff
@@ -113,13 +129,21 @@ async def _execute_async(self):
                 "min, stddev, stddev_pop, stddev_samp, sum, var_pop, var_samp, "
                 "variance."
             )
+        return {
```
small refactor: combine 2 methods

hotaru355 commented on cuckoo/finalizers.py:
```diff
@@ -183,9 +197,12 @@ def __init__(
+        invert_selection=False,
     ) -> Generator[TYIELD, None, None]:
-        self._node._fragments.response_keys.append(self._response_key(columns))
+        resolved_columns = self._resolve_column_selection(columns, invert_selection)
```
this adds the feature to each method

hotaru355 commented on cuckoo/finalizers.py:
```diff
@@ -183,9 +197,12 @@ def __init__(
     def yielding(
         self: YieldingFinalizer,
-        columns: TCOLUMNS = ["uuid"],
+        columns: Optional[TCOLUMNS] = None,
```
note that the default value is now dynamic - see comment above

hotaru355 commented on cuckoo/include.py:
```diff
@@ -20,18 +17,18 @@
         self,
         model: Type[TMODEL],
-        key: Optional[str] = None,
+        field_name: Optional[str] = None,
```
I renamed this argument, as I find the new name more suiting. this is a feature we do not use in any repo, so pls ignore.

Feature: if a model has 2 relations to the same sub-model, you need a way to tell cuckoo which of the 2 relations you actually want.

hotaru355 commented on cuckoo/include.py:
```diff
@@ -87,59 +84,50 @@ def bind_and_configure(parent: BinaryTreeNode):
 
         return AggregateIncludeFinalizer(node=self, finalize_fn=bind_and_configure)
 
-    def _get_relation_name(self, is_list=False):
```
OLD:
lots of pydantic V1 specific code baked into the `Include` model

hotaru355 commented on cuckoo/include.py:
```diff
@@ -87,59 +84,50 @@ def bind_and_configure(parent: BinaryTreeNode):
-                    lambda field: field[0] == self._key, parent_model.__fields__.items()
-                ),
-                (None, None),
+    def _get_relation_name(
```
NEW:
no pydantic specific code here, just calls to the abstract interface (more to this in the `TableModel`)

hotaru355 commented on cuckoo/models/builtins.py:
```diff
@@ -0,0 +1,124 @@
+from abc import ABC, abstractmethod
```
this is in essence the merged code from `models/aggregate.py` and `models/common.py`. changes highlighted below

hotaru355 commented on cuckoo/models/builtins.py:
```diff
@@ -0,0 +1,124 @@
+from pydantic.generics import GenericModel
+
+
+class TableModel(BaseModel, ABC):
```
The abstract base class of any model that wants to work with cuckoo. A model class just needs to implement the `fields` method that lists all fields of the class. all other methods of this class build on the single `fields` method.

hotaru355 commented on cuckoo/models/builtins.py:
```diff
@@ -0,0 +1,124 @@
+        extra = Extra.allow
+
+
+TMODEL = TypeVar("TMODEL", bound=TableModel)
+TBATCH_MODEL = TypeVar("TBATCH_MODEL", bound=TableModel)
+TNUM_PROPS = TypeVar("TNUM_PROPS")
+TMODEL_BASE = TypeVar("TMODEL_BASE")
+TBATCHNUM_PROPS = TypeVar("TBATCHNUM_PROPS")
+TBATCHMODEL_BASE = TypeVar("TBATCHMODEL_BASE")
+
+
+class Aggregate(GenericModel, Generic[TMODEL_BASE, TNUM_PROPS]):
+    count: Optional[int]
+    avg: Optional[TNUM_PROPS]
+    max: Optional[TMODEL_BASE]
+    min: Optional[TMODEL_BASE]
+    stddev: Optional[TNUM_PROPS]
+    stddev_pop: Optional[TNUM_PROPS]
+    stddev_samp: Optional[TNUM_PROPS]
+    sum: Optional[TNUM_PROPS]
+    var_pop: Optional[TNUM_PROPS]
+    var_samp: Optional[TNUM_PROPS]
+    variance: Optional[TNUM_PROPS]
+
+
+class AggregateResponse(GenericModel, Generic[TMODEL_BASE, TNUM_PROPS, TMODEL]):
+    aggregate: Optional[Aggregate[TMODEL_BASE, TNUM_PROPS]]
+    nodes: Optional[list[TMODEL]]
```
unchanged

hotaru355 commented on cuckoo/models/providers.py:
```diff
@@ -0,0 +1,49 @@
+from .builtins import TableModel
+
+
+class PydanticV1TableModel(TableModel):
```
pydantic V1 implementation of the abstract base class (it uses `__fields__` and `type_`, which are specific to pydantic V1)

hotaru355 commented on cuckoo/models/__init__.py:
```diff
@@ -1,9 +1,43 @@
+
+HasuraTableModel = PydanticV1TableModel
+"""Table model, defaults to pydantic V1. To use any other lib, use for example:
+`HasuraTableModel = PydanticV2TableModel`"""
```
To preserve backwards compatibility, the default table model is using pydantic V1

hotaru355 commented on tests/conftest.py:
```diff
@@ -1,18 +1,19 @@
+from httpx import AsyncClient, Client
+from pytest import fixture
 
+load_dotenv(".env.default")
```
the test container used to load the env file, but since we now run tests locally, we need to load the file in code

hotaru355 commented on tests/conftest.py:
```diff
@@ -21,19 +22,6 @@ def user_uuid():
-@fixture(scope="session")
-def event_loop():
-    """
-    REQUIRED to make pytest-asyncio plugin work with async sessions. We replace the
```
this is now "somewhat" fixed, as long as you add the `@mark.asyncio(scope="session")` annotation to every async test class. trying to just add the global `asyncio_mode = auto` flag will still produce this error...

hotaru355 commented on tests/fixture/mutation_fixture.py:
```diff
@@ -477,13 +480,8 @@ def run_mutation(_, __, BatchDelete: Type[BatchDelete], ___):
-                    "jsonb_dict",
-                ]
+                columns=[],
+                invert_selection=True,
```
new feature makes some tests easier :stuck_out_tongue:

hotaru355 commented on tests/test_query.py:
```diff
@@ -97,6 +103,66 @@ async def test_returning_all_fields_and_relations_with_conditions(
 
         assert_authors_ordered([actual_author], [expected_author])
 
+    @mark.parametrize(
```
test for this feature

hotaru355 reviewed (commented):
actual changes

clvcooke commented on tests/test_mutation.py:
```diff
@@ -406,8 +412,15 @@ async def test_returning_all_columns_and_relations(
+            author.copy(
+                update={
+                    "age": author.age + 1,
+                    "updated_by": user_uuid,
```
Confused why this got added?

hotaru355 replied:
sharp eyes! Before, tests that used the `all_columns()` helper function did not compare any `updated_at`, `updated_by`, etc (see this [change](https://github.com/hotaru355/cuckoo-hasura/pull/6/files#diff-02c3a3a504afaef7a801cd71c4511b14ea2ccdeb0d0c876249bf4335b2a8306dR370) ), as they seemed to just obfuscate the actual test. I changed that helper to really get _all_ columns to validate the new feature and that broke some test like this one.

clvcooke commented on tests/test_query.py:
```diff
@@ -97,6 +103,66 @@ async def test_returning_all_fields_and_relations_with_conditions(
+                lambda author: author.copy(exclude=AUTHOR_RELATIONS),
+            ),
+        ],
+        ids=[
```
Of this is nice, does this give you the id of the failed test case?

hotaru355 replied:
this is an optional argument for the `@parameterize` annotation. If you omit the this arg, the test cases are printed with the str value of the passed parameter - not very readable in case you have lambda functions or python objects as parameters:
```sh
test_some_parameterized_function[<object abc123>] - PASSED
test_some_parameterized_function[<object def456>] - PASSED
```

so, the `ids` make parameterized tests look prettier:
```sh
test_some_parameterized_function[with default value] - PASSED
test_some_parameterized_function[with complex value] - PASSED
```

clvcooke reviewed (commented):
Looking really good! I have the same question as @kaymanb that I'd like an answer to before merge

hotaru355 committed:
PR review fixes:
- added error handling when inverted columns are invalid
- added tests for error handling

hotaru355 commented on tests/test_query.py:
```diff
@@ -97,6 +107,108 @@ async def test_returning_all_fields_and_relations_with_conditions(
+                lambda author: author.copy(exclude=AUTHOR_RELATIONS),
+            ),
+            (
+                lambda: [Include(AuthorDetail).one().returning(invert_selection=True)],
```
Forgot to test the feature on included relations

hotaru355 commented on cuckoo/finalizers.py:
```diff
@@ -61,6 +64,25 @@ def __init__(
+            field_names = {field_name for field_name, _, _ in self._node.model.fields()}
+            invalid_columns = list(filter(lambda col: col not in field_names, columns))
+            if invalid_columns:
+                raise HasuraClientError(
```
raising an error for any invalid columns, but only when inverting the selection. validation of columns when _not_ using the feature would be more complex, especially validating a stringified relation like `articles { uuid }`. When you use the `Include` class, validation is done in that class.

hotaru355 reviewed (commented):
@clvcooke and @kaymanb 
Forgot to push the requested changes. Thanks!

hotaru355 committed:
version bump
